net-tools/netwatch/src/udp.rs

```diff
@@ -51,97 +68,810 @@ impl UdpSocket {
 }
 
-impl std::ops::Deref for UdpSocket {
-    type Target = tokio::net::UdpSocket;
+/// Send future
+#[derive(Debug)]
+pub struct SendToFut<'a, 'b> {
+    socket: &'b UdpSocket,
+    buffer: &'a [u8],
+    to: SocketAddr,
+}
 
-    fn deref(&self) -> &Self::Target {
-        self.0.as_ref().expect("only removed on drop")
+impl Future for SendToFut<'_, '_> {
+    type Output = std::io::Result<usize>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        loop {
+            // check if the socket needs a rebind
+            if self.socket.is_broken() {
+                match self.socket.rebind() {
+                    Ok(()) => {
+                        // all good
+                    }
+                    Err(err) => {
+                        warn!("failed to rebind socket: {:?}", err);
+                        // TODO: improve error
+                        let err =
+                            std::io::Error::new(std::io::ErrorKind::BrokenPipe, err.to_string());
+                        return Poll::Ready(Err(err));
+                    }
+                }
+            }
+
+            let guard = self.socket.socket.read().unwrap();
+            let Some((socket, _)) = guard.as_ref() else {
+                warn!("socket closed");
+                return Poll::Ready(Err(std::io::Error::new(
+                    std::io::ErrorKind::BrokenPipe,
+                    "socket closed",
+                )));
+            };
+
+            match socket.poll_send_ready(cx) {
+                Poll::Pending => {
+                    self.socket.send_waker.register(cx.waker());
+                    return Poll::Pending;
+                }
+                Poll::Ready(Ok(())) => {
+                    let res = socket.try_send_to(self.buffer, self.to);
+                    if let Err(err) = res {
+                        if err.kind() == ErrorKind::WouldBlock {
+                            continue;
+                        }
+
+                        if let Some(err) = self.socket.handle_write_error(err) {
+                            return Poll::Ready(Err(err));
+                        }
+                        continue;
+                    }
+                    return Poll::Ready(res);
+                }
+                Poll::Ready(Err(err)) => {
+                    if let Some(err) = self.socket.handle_write_error(err) {
+                        return Poll::Ready(Err(err));
+                    }
+                    continue;
+                }
+            }
+        }
     }
 }
 
+fn inner_bind(addr: SocketAddr) -> Result<(tokio::net::UdpSocket, quinn_udp::UdpSocketState)> {
+    let network = IpFamily::from(addr.ip());
+    let socket = socket2::Socket::new(
+        network.into(),
+        socket2::Type::DGRAM,
+        Some(socket2::Protocol::UDP),
+    )
+    .context("socket create")?;
+
+    if let Err(err) = socket.set_recv_buffer_size(SOCKET_BUFFER_SIZE) {
+        warn!(
```

Comment:
We need to downgrade these to debug.  We've had reports that this fails on iOS.  (Sadly not filed as an issue.)  And these logs don't just appear once but re-appear on every rebind.

Reply:
done